File: VB/App_Code/Operation.vb
Option Infer On

Imports System
Imports System.Threading

Public Class Operation
    Private _progress As Int32
    Private _started As Boolean
    Private l_started As New Object()

    Public Property Progress() As Int32
        Get
            Return _progress
        End Get
        Private Set(ByVal value As Int32)
            _progress = value
        End Set
    End Property

    Public Property Started() As Boolean
        Get
            SyncLock l_started
                Return _started
            End SyncLock
        End Get
        Private Set(ByVal value As Boolean)
            SyncLock l_started
                _started = value
            End SyncLock
        End Set
    End Property

    Public Sub AsyncStart(ByVal interval As Int32)
        If Started Then
            Return
        End If
        Started = True
        Progress = 0
        Dim t = New Thread(AddressOf start)
        t.Start(interval)
    End Sub

    Private Sub start(ByVal interval As Object)
'        
'         * suggest long operations here
'         
        Dim time = DirectCast(interval, Int32)
        Const steps As Int32 = 100
        Dim period = time \ steps
        For i As Integer = 0 To steps - 1
            Thread.Sleep(period)
            Progress += 1
        Next i
        Started = False
    End Sub
End Class

File: CS/App_Code/Operation.cs


using System;
using System.Threading;

public class Operation {
    Int32 _progress;
    Boolean _started;
    Object l_started = new Object();

    public Int32 Progress {
        get { return _progress; }
        private set { _progress = value; }
    }

    public Boolean Started {
        get { lock (l_started) return _started; }
        private set { lock (l_started) _started = value; }
    }

    public void AsyncStart(Int32 interval) {
        if (Started) return;
        Started = true;
        Progress = 0;
        var t = new Thread(start);
        t.Start(interval);
    }

    void start(Object interval) {
        /*
         * suggest long operations here
         */
        var time = (Int32)interval;
        const Int32 steps = 100;
        var period = time / steps;
        for (int i = 0; i < steps; i++) {
            Thread.Sleep(period);
            Progress++;
        }
        Started = false;
    }
}

